### ISSUE TITLE
ERL-919: Sampling profiler -- erlang:statistics(currently_running)?

### ISSUE THREAD

**Issue by OTP-Maintainer** [erlang/otp#3783]
Original reporter: `bucko909`
Affected version: `Not Specified`
Component: `Not Specified`
Migrated from: https://bugs.erlang.org/browse/ERL-919

---

```
Hi! I'd like to do some sample-based profiling of our live system to simply find resource hogs at busy times. The best we can currently do at the moment is something like:

{noformat}
[
Stack ||
  Pid <- erlang:processes(),
  Pid /= self(),
  [{current_stacktrace, Stack}, {status, running}] <-
    [erlang:process_info(Pid, [current_stacktrace, status])
]
{noformat}

(Forgive my inability to format in JIRA.)

This is suboptimal for a few reasons:

* It can't pick up very short-lived processes (which die or are created between {{processes()}} and {{process_info()}}), which gives a sampling bias towards longer-lived processes. We've found at times that much of our load comes from millions of spawned workers.
* It's massively inefficient and at busy times, extremely slow, taking almost a second even with just 10k processes. I only need at most 16 samples with my 16 cores, not 1000 times that.
* I'm not _sure_ what happens when you get the stack trace of a process in "running" state but my suspicion, as the code appears to involve locks and signals, is that there's an inherent bias towards certain "safe" points.

I can also turn on full tracing of the entire system with trace-provided timestamps, then pick sampling points along the timeline, but:
* This is really super-duper heavyweight and clearly affects the timing characteristics of the runtime.
* It only tells me what's _runnable_, not what's _running_. At busy times, anything that's runnable will be overestimated in duration as it's being perpetually paused by the BEAM to make way for everything else.
* (The {{running}} or {{running_procs}} trace options are better, but can't tell me stacktraces. Polling for stacks gives the same biases as the {{processes()}} approach.)
* This method doesn't tell me how processes migrate between schedulers.
* Using the calls consumes the tracer resource, meaning no other tracing can be done.

There's also currently a call {{erlang:statistics(run_queue_lengths)}} which includes an item for each scheduler indicating its run queue length, but not what's in it!

I'd like some way of writing a lightweight sampling profiler taking, say, a hundred samples per second of only running processes with minimal (or preferably no) locks. I propose adding some new {{erlang:statistics/1}} calls which return a list in the same way as, say, {{erlang:statistics(run_queue_lengths)}}:

* {{currently_running}} -- returns {noformat}[pid() | port()]{noformat} (or anything else a scheduler can be doing). This is effectively the same as tracing {{running_procs}} but costs nothing if you never ask for the data, and gives per-scheduler stats.
* {{currently_running_mfas}} -- returns {noformat}[{pid(), mfa()} | {port(), undefined}]{noformat}
* {{currently_running_stacktraces}} -- returns {noformat}[{pid(), stacktrace()} | {port(), undefined}]{noformat}

It might also be beneficial to have {{currently_runnable}} varieties of the above indicating what work is scheduled (as opposed to actually executing). This data will tend to have a bias towards things like {{gen_server:loop}} (processes which have a message on their queue but haven't started processing it yet), making it less valuable, of course.

In principle this doesn't seem _much_ harder than the current counting varieties of the same calls.

I believe these features will:
* Allow some rather powerful statistical profiling techniques.
* Allow developers to more easily diagnose and understand issues with scheduler behaviours.

A disadvantage I can think of easily is that, of course, the process making the call is itself consuming a scheduler and will bias samples (anything residing on that scheduler will never show up as running, and it'll appear as using 100% of a CPU even sampling only once per second!); perhaps some async version could fix that, but I think that doesn't matter too much for now: I'd rather see good stats for every scheduler but my own than have no stats at all. Each call returns {{pid()}} so finding my own scheduler is pretty easy!

The thing I most care about is getting {{currently_running}} and {{currently_runnable}} working. The rest can come later if there's difficulties, or be implemented with a known sampling bias.

I'm prepared to have a go myself, but I've no idea how to go about actually reading the current process out of a scheduler in a safe way -- I can see a {{current_process}} pointer, which I can atomically read, but how can I ensure it's not released by the scheduler after reading it?
```

**Comment by max-au** [comment]
Bumped into this one. With BeamAsm it seems no longer needed...

**Comment by garazdawi** [comment]
I would say that it is still needed on non-linux OSs and in places where you cannot use `lbr`.

Though the possibility to use `perf` to record stack of erlang process does make this less needed.
